pkg/restore/restore.go

```diff
@@ -510,6 +501,116 @@ func (ctx *context) restoreFromDir(dir string) (api.RestoreResult, api.RestoreRe
+func ensureNamespaceIsReady(ns *v1.Namespace, logger *logrus.Entry, errs *api.RestoreResult, nsClient corev1.NamespaceInterface) wait.ConditionFunc {
+	return func() (bool, error) {
+		clusterNS, err := nsClient.Get(ns.Name, metav1.GetOptions{})
+		if apierrors.IsNotFound(err) {
+			if _, err := nsClient.Create(ns); err != nil {
+				addArkError(errs, err)
+				return false, err
+			}
+
+			// ns created successfully; exit the loop
+			logger.Infof("Termination successful, namespace %s restored", ns.Name)
+			return true, nil
+		} else if err != nil {
+			addArkError(errs, err)
+			return false, err
+		}
+
+		if clusterNS != nil && clusterNS.Status.Phase == v1.NamespaceTerminating {
+			// keep looping
+			logger.Debugf("Namespace %s still terminating, waiting", ns.Name)
+			return false, nil
+		}
+
+		// namespace exists and is not terminating; exit the loop
+		return true, nil
+	}
+}
+
+// ensurePVReadyToCreate creates a wait.ConditionFunc for controlling a wait.PollImmediate loop.
+// The returned function will break the loop if:
+//		- the target persistent volume exists and does not have a DeletionTimestamp set.
+//		- the target persistent volume is not found
+//		- the target persistent volume has a PVC that's associated with a namespace, and looking up the namespace fails
+// If the target persistent volume is not found, a new persistent disk is created and the object-to-restore updated to reference the new disk. Otherwise, a new persistent disk is not made.
+func (ctx *context) ensurePVReadyToCreate(obj *unstructured.Unstructured, errs *api.RestoreResult, warnings *api.RestoreResult, resourceClient client.Dynamic, fullPath string) wait.ConditionFunc {
+	return func() (bool, error) {
+
+		name := obj.GetName()
+		clusterPV, err := resourceClient.Get(name, metav1.GetOptions{})
+		// Check if the PV exists in the cluster before attempting to create
+		// a volume from the snapshot, in order to avoid orphaned volumes (GH #609)
+		// PV's existence will be recorded later. Just skip the volume restore logic.
+		if apierrors.IsNotFound(err) {
+			ctx.log.Infof("Restoring  PersistentVolume %s", name)
+			// restore the PV from snapshot (if applicable)
+			updatedObj, err := ctx.pvRestorer.executePVAction(obj)
+			if err != nil {
+				addToResult(errs, "", fmt.Errorf("error executing PVAction for %s: %v", fullPath, err))
+				return false, err
+			}
+
+			// Copy the new disks's information into the PV we're trying to restore
+			updatedObj.DeepCopyInto(obj)
+
+			return true, nil
+		} else if err != nil {
+			addToResult(errs, "", fmt.Errorf("error checking existence for PV %s: %v", name, err))
+			return false, err
+		}
+
+		if clusterPV != nil && clusterPV.GetDeletionTimestamp() == nil {
+			namespace, err := collections.GetString(clusterPV.UnstructuredContent(), "spec.claimRef.namespace")
+
+			// The PV is likely missing a claimRef, Assume that this means we're waiting for a namespace to delete.
+			if err != nil {
```

Comment:
I'd rather be certain as to why we're getting an error - I'd suggest either using `collections.Exists` first, or converting to a `v1.PersistentVolume`